src/cascadia/TerminalApp/AppKeyBindings.cpp

```diff
@@ -1,117 +1,294 @@
-                _CloseWindowHandlers();
-                return true;
-            case ShortcutAction::CloseTab:
-                _CloseTabHandlers();
-                return true;
-
-            case ShortcutAction::ScrollUp:
-                _ScrollUpHandlers();
-                return true;
-            case ShortcutAction::ScrollDown:
-                _ScrollDownHandlers();
-                return true;
-
-            case ShortcutAction::NextTab:
-                _NextTabHandlers();
-                return true;
-            case ShortcutAction::PrevTab:
-                _PrevTabHandlers();
-                return true;
-        }
-        return false;
-    }
-
-    // -------------------------------- Events ---------------------------------
-    DEFINE_EVENT(AppKeyBindings, CopyText,          _CopyTextHandlers,          TerminalApp::CopyTextEventArgs);
-    DEFINE_EVENT(AppKeyBindings, PasteText,         _PasteTextHandlers,         TerminalApp::PasteTextEventArgs);
-    DEFINE_EVENT(AppKeyBindings, NewTab,            _NewTabHandlers,            TerminalApp::NewTabEventArgs);
-    DEFINE_EVENT(AppKeyBindings, NewTabWithProfile, _NewTabWithProfileHandlers, TerminalApp::NewTabWithProfileEventArgs);
-    DEFINE_EVENT(AppKeyBindings, NewWindow,         _NewWindowHandlers,         TerminalApp::NewWindowEventArgs);
-    DEFINE_EVENT(AppKeyBindings, CloseWindow,       _CloseWindowHandlers,       TerminalApp::CloseWindowEventArgs);
-    DEFINE_EVENT(AppKeyBindings, CloseTab,          _CloseTabHandlers,          TerminalApp::CloseTabEventArgs);
-    DEFINE_EVENT(AppKeyBindings, SwitchToTab,       _SwitchToTabHandlers,       TerminalApp::SwitchToTabEventArgs);
-    DEFINE_EVENT(AppKeyBindings, NextTab,           _NextTabHandlers,           TerminalApp::NextTabEventArgs);
-    DEFINE_EVENT(AppKeyBindings, PrevTab,           _PrevTabHandlers,           TerminalApp::PrevTabEventArgs);
-    DEFINE_EVENT(AppKeyBindings, IncreaseFontSize,  _IncreaseFontSizeHandlers,  TerminalApp::IncreaseFontSizeEventArgs);
-    DEFINE_EVENT(AppKeyBindings, DecreaseFontSize,  _DecreaseFontSizeHandlers,  TerminalApp::DecreaseFontSizeEventArgs);
-    DEFINE_EVENT(AppKeyBindings, ScrollUp,          _ScrollUpHandlers,          TerminalApp::ScrollUpEventArgs);
-    DEFINE_EVENT(AppKeyBindings, ScrollDown,        _ScrollDownHandlers,        TerminalApp::ScrollDownEventArgs);
-
-
-}
+﻿// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+#include "pch.h"
+#include "AppKeyBindings.h"
+
+using namespace winrt::Microsoft::Terminal;
+using namespace winrt::TerminalApp;
+using namespace winrt::Windows::Data::Json;
+
+static const std::wstring KEYS_KEY{ L"keys" };
+static const std::wstring COMMAND_KEY{ L"command" };
+
+static const std::wstring COPYTEXT_KEY{ L"copy" };
+static const std::wstring PASTETEXT_KEY{ L"paste" };
+static const std::wstring NEWTAB_KEY{ L"newTab" };
+static const std::wstring NEWTABWITHPROFILE0_KEY{ L"newTabProfile0" };

```

Comment:
i'm concerned about the explosion here. should key bindings be more like .. functions with parameters? `newTab(0)`, `newTab()`?

Reply:
if we hardcode function-like names today, we can even get a bit of compatibility if we ever move to function style

but at the same time... perhaps we don't want to overdesign

Reply:
It works well for `scrollUp`, too, so people can have custom deltas.

Reply:
See, this is something that I think is a good idea, but I thought it was maybe out of scope for this PR.
I was thinking it'd be something like the following:
```json
	{ "keys": ["ctrl+up"], "command": "scroll_lines", "args": {"amount": 1.0 } },
	{ "keys": ["ctrl+down"], "command": "scroll_lines", "args": {"amount": -1.0 } },
```
where then `args` is just parsed as a JsonObject by whatever ShortcutAction needs it. How that ShortcutAction actually identifies that it wants to be able to parse args, and how it registers it's parsing/serializing callback isn't something that I've figured out quite yet.

Reply:
clever. i like it

Reply:
@zadjii-msft Quick question - For the std::wstring constants - Couldn't these be const wchar_t pointers? (or plain wchar_t arrays) Like
`const wchar_t* const COPYTEXT_KEY = L"copy";
 `
This avoids static storage duration for these variables.   Not a huge deal with this code but it is a good guideline across a codebase.  This avoids potential lifetime issues at shutdown.

Reply:
I'm gonna summon @adiviness who's our resident code standards/guidlines guru to answer that question

Reply:
Alternatively, you could use constexpr.  
`constexpr wchar_t COPYTEXT_KEY[] = L"copy"; `

Reply:
In my opinion they should be `static constexpr std::wstring`.

Reply:
@adiviness looks like that fails with `'KEYS_KEY': illegal initialization of 'constexpr' entity with a non-constant expression`.

Reply:
Thanks, @adiviness.   I don't think that will work.  std::wstring has a non-trivial destructor.  Using the wchar_t array here is a drop in replacement for the std::wstring.  (I am assuming this by looking at the code)   

We follow this [guideline](https://google.github.io/styleguide/cppguide.html#Static_and_Global_Variables)  because we have been hit with the hard to find bugs that happen with static variables.   Again - this case looks harmless so not a huge deal, but it is a good practice to follow.